### ISSUE TITLE
Auto move

### ISSUE THREAD

**Issue by yephny** [w2sv/FileNavigator#14]
Hi, I think the app is great and fills a serious gap. I noticed in the description the mention of an auto move feature, but I can't see how I can enable the feature.

**Comment by w2sv** [comment]
Hey, thanks a lot! The 'auto move' feature refers to the one, that enables you to move a file to a previously selected destination via the `To /{previous_destination_dir_name}` notification action, without having to manually select the directory once again. The feature name might admittedly be a little misleading, you're right on that. I'll try to think of a better one.

How exactly did you expect the auto move feature to work?

**Comment by yephny** [comment]
I thought there would be a way to set a directory for each file type so that the app would automatically move downloaded files without user interaction.

**Comment by w2sv** [comment]
Okay I see, should be feasible and certainly would be a nice addition to the app functionality. I'll try to implement it asap.

**Comment by yephny** [comment]
The app is great already the way it is, for sure it'd be even better with this feature implemented. File management, or lack thereof, has been one of my biggest gripes with Android. This app solves it.

**Comment by b02860de585071a2** [comment]
I also assumed that the intention of "auto save" was to filter and automatically move files without user interaction. To my knowledge, this functionality isn't provided by any application currently available (at least on F-Droid or in the FOSS sphere of influence).

Some kind of regex filtering system that acts on user-defined and nested rules would be AMAZING, I'm not sure how much of a battery/CPU impact it would have though? Might need to be something that runs on a fixed schedule or watches for filesystem changes.


Example use cases:
- Any files found in `/[...]/BadFolder/` are moved to `/[...]/GoodFolder/` _(useful for apps that download/store files in annoying or unwanted locations)_
- Files of type `image` created in `/[...]/Download/` OR `/[...]/Documents/ExampleSubfolder/` are moved to `/[...]/Pictures/`
- Files that match the regex `.*\.apk` are moved to `/[...]/APK/`, then anything in that folder matching `(?i).*github.*` is further moved to `/[...]/APK/GitHub/`


I have written some scripts for my Linux machines that perform similar tasks, having this on Android too in a somewhat "official" manner (instead of creating Termux scripts, etc) would be absolutely fantastic.

**Comment by deathblade666** [comment]
I agree the name auto move would imply no user interaction at all. I have only seen one other app that does this and it hasnt been updated in nearly a decade and assumed discontinued (no longer on play store, and sadly not FOSS) called redirect pro. I think your app has ALOT of potential and could be the perfect tool I've been looking for if it had a true auto move.

Thanks a TON for creating such an amazing app thus far, can't wait to see whats in store for it in future releases!!

**Comment by w2sv** [comment]
> Yes, I also was kinda misled by the name "auto move" 😄
> 
> I hope this will be implemented properly soon, as well as adding more file extensions such as `.epub` and `.mobi`.
> 
> This project has a lot of potential. Great work!

An eBook file type is a great idea which really ain't hard to implement. Adding more file types is a breeze in general, so don't hold back on suggestions on that end :) 
I'll create a separate issue for eBooks in any case, thank you!

**Comment by w2sv** [comment]
> Some kind of regex filtering system that acts on user-defined and nested rules would be AMAZING, I'm not sure how much of a battery/CPU impact it would have though? Might need to be something that runs on a fixed schedule or watches for filesystem changes.

The file detection mechanism that triggers move notifications is implemented through [ContentObserver](https://developer.android.com/reference/android/database/ContentObserver)s which are registered for external content URIs and trigger a callback, whenever a file corresponding to the respective content URI is changed. Therefore, the battery impact is really minimal, as its a reactive mechanism, however, it is also why non-media (image, video, audio) file types can't be discovered outside of the system-specific Download folder, as the Android SDK only provides external content URIs for the media types image, video, audio and download.

**Comment by w2sv** [comment]
In any case I've started working on the actual auto move feature. I thought I'll do it so that one can determine the auto move destination for a file type as a whole, as well as for the respective file type sources, which override the file type destination if provided. The UI integration into the settings screen still causes me some headaches, but I'll figure it out.

Regex-based filtering is a great idea which I'll keep in mind for a later release too, however for now I'll focus on making a 'basic' auto move mechanism work properly. Will be complicated and issue-prone enough in and itself I believe :D

**Comment by b02860de585071a2** [comment]
> Regex-based filtering is a great idea which I'll keep in mind for a later release too, however for now I'll focus on making a 'basic' auto move mechanism work properly. Will be complicated and issue-prone enough in and itself I believe :D

Anything is better than what's currently available on the market (ie, nothing lol)! 

Lately I've been fighting with working around the limited binaries in the adb shell/ksh, trust me I get it. Android development is weirdly hard.

**Comment by w2sv** [comment]
![Screenshot_2024-06-06-19-34-11-912_com w2sv filenavigator debug](https://github.com/w2sv/FileNavigator/assets/36566555/04fce194-5ef6-4869-9095-556afe99237c)

Just wanted to let you know that I've come to terms with what I'd consider a decent UI implementation, still got a bit of a way to go until the new release though.

**Comment by w2sv** [comment]
> Android development is weirdly hard.

True dat ;)

**Comment by b02860de585071a2** [comment]
> ![Screenshot_2024-06-06-19-34-11-912_com w2sv filenavigator debug](https://github.com/w2sv/FileNavigator/assets/36566555/04fce194-5ef6-4869-9095-556afe99237c)
> 
> Just wanted to let you know that I've come to terms with what I'd consider a decent UI implementation, still got a bit of a way to go until the new release though.


Love it!

Just out of curiosity: is this acting on MIME types, file extensions, or something else? I was reading some documentation and it seems like Android MIME is not as fully fleshed out as *nix, despite sharing many similarities.

**Comment by w2sv** [comment]
> Just out of curiosity: is this acting on MIME types, file extensions, or something else? I was reading some documentation and it seems like Android MIME is not as fully fleshed out as *nix, despite sharing many similarities.

A mix of both. There are 4 different types of `ContentObserver`s, one respectively for Image, Video, Audio and Download, which are registered or not as per the user file type settings. I'm not quite sure how Android determines which one to call upon a new file entering the file system, but I'm guessing it'll be based on MIME types. So if the Image, Video or Audio `ContentObserver` is called, I know that it's a file of that respective type, upon the Download one being triggered, I'm determining the file type through file extensions I [defined myself](https://github.com/w2sv/FileNavigator/blob/2443a935f246dcf1c6d87ba51ce60b9d9afc9ecc/core/domain/src/main/kotlin/com/w2sv/domain/model/FileType.kt#L122). 
The source type, on the other hand, is identified by the [directory the file is found in](https://github.com/w2sv/FileNavigator/blob/2443a935f246dcf1c6d87ba51ce60b9d9afc9ecc/core/navigator/src/main/kotlin/com/w2sv/navigator/mediastore/MediaStoreColumnData.kt#L73).

**Comment by w2sv** [comment]
Auto moving is now possible in [0.2.0](https://github.com/w2sv/FileNavigator/releases/tag/0.2.0) btw, let me know what y'all think :)

**Comment by deathblade666** [comment]
Seems to work Great! Only thing I'd ask in addition is an exclude specific folder option per file type. If that's already added then ignore this. Haven't had time to comb through everything yet

**Comment by boognish-rising** [comment]
Hey there! It looks like I'm late to the party but I just noticed the Auto Move update. I'll admit that in its absence (when I fully thought it was effectively what the app was all about), I hadn't gotten much use out of FN yet but I saw potential so I kept it installed... Well, I'm glad I did because it appears the app's trajectory is aligned with what I had previously thought. My reason for barging in on the discussion today, though, is to ask if the feature is limited to files of X type being transferred automatically (or following a notification/prompt, as configured is the Navigator Settings) from the time the Auto Move destination/link is created onward and if so, whether you'd consider implementing the transfer of existing files of X type once the destination is set. 

I could see users responding to this behavior with support for & against, so it'd probably be something best left configurable/with a simple toggle to set default behavior, but I know that I for one would very much appreciate if you extended the functionality of Auto Move to include it! 

PS - while more of an afterthought than it is in File Navigator, X-plore File Manager has a feature that's at least similar _in concept_ to Auto Move (I think it's just called File Sync). In fact,_on paper/in theory_ it's exactly what I'm looking for, i.e., both pre-existing files as well as any added from then on are to be transferred (navigated) accordingly, however, it seems to be bug-ridden. The link between source & destination is inexplicably lost and results in it being anything but automatic/automated (once they're no longer connected/synced). I set up 2 separate devices to transfer files from A->B (my PC via network share). Both worked as expected for a bit, but before long both became disconnected and the effort put into setting them up - wasted. All that to say, Android users could benefit a great deal from a user-friendly (opposed to apps like Syncthing, which is great but def more suited for advanced/power users), functional app with such capabilities!

**Comment by JaneValleyhasBtaken** [comment]
Is it possible that move all files existed before installation of the app to designated folders? The app concept is great. However, my file is already a mess and I really don't want to do it myself :(.

**Comment by b02860de585071a2** [comment]
New feature is great! Almost exactly what I'm looking for. 

I do agree that acting on existing files would really make it shine.

**Comment by HasanAbbadi** [comment]
> Is it possible that move all files existed before installation of the app to designated folders? The app concept is great. However, my file is already a mess and I really don't want to do it myself :(.

For the time being, you can use a file manager like X-Plore. Just hit the search icon and search for all files of a certain type using regex, like: `*.pdf`. Then, select all and press move. This will move them to a folder that you opened on the other pane.

While it is not painless, it's surely much quicker than individually looking for files or waiting for this feature to be implemented :)
